test/broadcasting.jl

```diff
@@ -0,0 +1,364 @@
+    @test_throws DimensionMismatch df[:, [:x1,:x2]] .= rand(3, 10)
+    @test_throws DimensionMismatch dfv[:, [:x3,:x4]] .= rand(2, 10)
+end
+
+@testset "assignment to a whole data frame and data frame row" begin
+    df = copy(refdf)
+    df .= 10
+    @test all(Matrix(df) .== 10)
+    dfv = view(df, 1:2, 1:4)
+    dfv .= 100
+    @test (all(Matrix(df[1:2, 1:4]) .== 100))
+    @test (all(Vector(df[3, 1:4]) .== 10))
+    dfr = df[1, 1:2]
+    dfr .= 1000
+    @test (all(Vector(df[1, 1:2]) .== 1000))
+    @test (all(Vector(df[3, :]) .!= 1000))
+
+    df = copy(refdf)
+    df[:] .= 10
+    @test all(Matrix(df) .== 10)
+    dfv = view(df, 1:2, 1:4)
+    dfv[:] .= 100
+    @test (all(Matrix(df[1:2, 1:4]) .== 100))
+    @test (all(Vector(df[3, 1:4]) .== 10))
+    dfr = df[1, 1:2]
+    dfr[:] .= 1000
+    @test (all(Vector(df[1, 1:2]) .== 1000))
+    @test (all(Vector(df[3, :]) .!= 1000))
+
+    df = copy(refdf)
+    df[:,:] .= 10
+    @test all(Matrix(df) .== 10)
+    dfv = view(df, 1:2, 1:4)
+    dfv[:, :] .= 100
+    @test (all(Matrix(df[1:2, 1:4]) .== 100))
+    @test (all(Vector(df[3, 1:4]) .== 10))
+end
+
+@testset "extending data frame in broadcasted assignment - one column" begin
+    df = copy(refdf)
+    df[:a] .= 1
+    @test all(Matrix(df) .== 1)
+    @test names(df)[end] == :a
+    df[:b] .= [1, 1, 1]
+    @test all(Matrix(df) .== 1)
+    @test names(df)[end] == :b
+    cdf = copy(df)
+    @test_throws DimensionMismatch df[:c] .= ones(3, 1)
+    @test df == cdf
+    @test_throws DimensionMismatch df[:x] .= ones(4)
+    @test df == cdf
+    @test_throws BoundsError df[10] .= ones(3)
+    @test df == cdf
+
+    dfv = @view df[1:2, 2:end]
+    @test_throws BoundsError dfv[10] .= ones(3)
+    @test_throws ArgumentError dfv[:z] .= ones(3)
+    @test df == cdf
+    dfr = df[1, 3:end]
+    @test_throws BoundsError dfr[10] .= ones(3)
+    @test_throws ArgumentError dfr[:z] .= ones(3)
+    @test df == cdf
+end
+
+@testset "empty data frame corner case" begin
+    df = DataFrame()
+    @test_throws ArgumentError df[1] .= 1
+    @test_throws ArgumentError df[:a] .= [1]
+    @test_throws ArgumentError df[[:a,:b]] .= [1]
+    @test df == DataFrame()
+    df .= 1
+    @test df == DataFrame()
+    df .= [1]
+    @test df == DataFrame()
+    df .= ones(1,1)
+    @test df == DataFrame()
+    @test_throws DimensionMismatch df .= ones(1,2)
+    @test_throws DimensionMismatch df .= ones(1,1,1)
+
+    @test_throws ArgumentError df[:a] .= 1
+    @test_throws ArgumentError df[[:a, :b]] .= 1
+
+    df = DataFrame(a=[])
+    @test_throws ArgumentError df[:b] .= 1
+end
+
+@testset "test categorical values" begin
+    df = copy(refdf)
+    v = categorical([1,2,3])
+    df[:c1] .= v
+    @test df.c1 == v
+    @test df.c1 !== v
+    @test df.c1 isa CategoricalVector
+    @test levels(df.c1) == levels(v)
+    @test levels(df.c1) !== levels(v)
+    df[:c2] .= v[1]
+    @test df.c2 == [1,1,1]
+    @test df.c2 isa CategoricalVector
+    @test levels(df.c2) != levels(v)
```

Comment:
This is somewhat unfortunate. We should allow `allocatecolumn` to take a value rather than just a type so that levels can be preserved (not for this PR).

Reply:
I have opened https://github.com/JuliaData/Tables.jl/issues/99 to keep track of it

Reply:
Why is this unfortunate? Does having consistent levels really matter? I'm not against adjusting allocatecolumn, but maybe we should also figure out a more general solution in DataAPI.jl that allows other array types a similar kind of flexibility.

Reply:
It's unfortunate because the whole point of `CategoricalValue` is to preserve custom ordering of levels. If we allowed `allocatecolumn` to take values and not just types, any array type could reuse that mechanism if needed.

Reply:
My thinking is similar to @nalimilan. The whole design idea of `CategoricalValue` is that it carries over the level information if I understand the original intention of @nalimilan correctly here.

If someone does not need this (ignoring other differences in functionalities) then probably `PooledArray` is what should be used.

Regarding the generality required. Currently there are no important types that carry over metadata AFAIK, but e.g. in the future we might have metadata attached to a `DataFrame` object.